File: 3887-minimum-cost-path-with-edge-reversals/minimum-cost-path-with-edge-reversals.py
import heapq as h 

class Solution:
    def minCost(self, n: int, edges: List[List[int]]) -> int:

        adj  = defaultdict(list)
        indegree = defaultdict(list)

        #
        for a,b,w in edges:
            adj[a].append((b,w))
            indegree[b].append((a,w))

        min_heap =[(0,0)]
        d = [ float("inf") for i in range(n)]
        d[0] = 0
        while(min_heap):
            #print( min_heap)
            node_wt, node = h.heappop(min_heap)
            if node  == n-1:
                return d[node]
            for nb,nb_wt in adj[node]:
                if node_wt + nb_wt < d[nb]:
                    h.heappush(min_heap,( node_wt + nb_wt,nb))
                    d[nb] = node_wt +nb_wt

            for nb,nb_wt in indegree[node]:
                if node_wt + 2*nb_wt < d[nb]:
                    h.heappush(min_heap,( node_wt + 2*nb_wt, nb))
                    d[nb] = node_wt +2*nb_wt
            


        return -1
                

File: 3887-minimum-cost-path-with-edge-reversals/minimum-cost-path-with-edge-reversals.java
import java.util.*;

class Pair {
    int dist;
    int node;

    Pair(int dist, int node) {
        this.dist = dist;
        this.node = node;
    }
}

class Edge {
    int to;
    int wt;

    Edge(int to, int wt) {
        this.to = to;
        this.wt = wt;
    }
}

class Solution {
    public int minCost(int n, int[][] edges) {

        Map<Integer, List<Edge>> adj = new HashMap<>();
        Map<Integer, List<Edge>> indegree = new HashMap<>();

        // Build graph
        for (int[] e : edges) {
            int a = e[0], b = e[1], w = e[2];

            adj.computeIfAbsent(a, k -> new ArrayList<>())
               .add(new Edge(b, w));

            indegree.computeIfAbsent(b, k -> new ArrayList<>())
                    .add(new Edge(a, 2 * w));
        }

        int[] d = new int[n];
        Arrays.fill(d, Integer.MAX_VALUE);
        d[0] = 0;

        PriorityQueue<Pair> pq =
                new PriorityQueue<>((x, y) -> x.dist - y.dist);

        pq.add(new Pair(0, 0));

        while (!pq.isEmpty()) {
            Pair cur = pq.poll();
            int node = cur.node;
            int nodeWt = cur.dist;

            if (node == n - 1)
                return d[node];

       

            // normal outgoing edges
            if (adj.containsKey(node)) {
                for (Edge e : adj.get(node)) {
                    if (nodeWt + e.wt < d[e.to]) {
                        d[e.to] = nodeWt + e.wt;
                        pq.add(new Pair(d[e.to], e.to));
                    }
                }
            }

            // reversed incoming edges
            if (indegree.containsKey(node)) {
                for (Edge e : indegree.get(node)) {
                    if (nodeWt + e.wt < d[e.to]) {
                        d[e.to] = nodeWt + e.wt;
                        pq.add(new Pair(d[e.to], e.to));
                    }
                }
            }
        }

        return -1;
    }
}
